sql/rowenc: reduce index key prefix calls

annrpom opened the pull request:
### sql/rowexec: run BenchmarkIndexBackfill on-disk
Epic: none
Release note: None

---


### sql/rowenc: reduce index key prefix calls
This patch removes redundant calls to `MakeIndexKeyPrefix` during
the construction of `IndexEntry`s by saving each first-time call in a
map that we can later lookup. Previously, we would make this call
for each row; however, as the prefix (table id + index id) for a
particular index remains the same, we do not need to do any
recomputation.

Epic: [CRDB-42901](https://cockroachlabs.atlassian.net/browse/CRDB-42901)
Fixes: https://github.com/cockroachdb/cockroach/issues/137798

Release note: None

```
name                             old time/op    new time/op    delta
IndexBackfill/100_rows-10          21.1ms ± 2%    21.4ms ± 4%    ~     (p=0.173 n=8+10)
IndexBackfill/10,000_rows-10       38.1ms ± 3%    38.1ms ± 6%    ~     (p=0.696 n=8+10)
IndexBackfill/1,000,000_rows-10     10.9s ± 5%     10.7s ± 1%    ~     (p=0.278 n=10+9)

name                             old alloc/op   new alloc/op   delta
IndexBackfill/100_rows-10          4.52MB ± 3%    4.81MB ±20%    ~     (p=0.053 n=9+10)
IndexBackfill/10,000_rows-10       25.8MB ± 8%    24.9MB ± 3%    ~     (p=0.156 n=10+9)
IndexBackfill/1,000,000_rows-10    6.25GB ± 3%    6.21GB ± 3%    ~     (p=0.353 n=10+10)

name                             old allocs/op  new allocs/op  delta
IndexBackfill/10,000_rows-10         120k ± 2%      109k ± 2%  -9.40%  (p=0.000 n=10+10)
IndexBackfill/1,000,000_rows-10     26.9M ± 0%     25.8M ± 1%  -3.84%  (p=0.000 n=8+10)
IndexBackfill/100_rows-10           24.0k ± 1%     24.7k ± 8%    ~     (p=0.156 n=9+10)
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139256)
<!-- Reviewable:end -->

annrpom committed:
sql/rowexec: run BenchmarkIndexBackfill on-disk

Epic: none
Release note: None

annrpom commented:
using benchdiff:
```
name                             old time/op    new time/op    delta
IndexBackfill/100_rows-10          40.8ms ± 4%    41.0ms ± 4%    ~     (p=0.340 n=9+9)
IndexBackfill/10,000_rows-10       47.8ms ± 4%    47.3ms ± 2%    ~     (p=0.400 n=10+9)
IndexBackfill/1,000,000_rows-10     10.9s ±10%     10.9s ± 1%    ~     (p=0.270 n=10+7)

name                             old alloc/op   new alloc/op   delta
IndexBackfill/100_rows-10          7.70MB ±10%    7.26MB ± 7%  -5.74%  (p=0.043 n=10+10)
IndexBackfill/10,000_rows-10       31.9MB ± 4%    32.0MB ± 9%    ~     (p=1.000 n=10+10)
IndexBackfill/1,000,000_rows-10    6.30GB ± 3%    6.28GB ± 3%    ~     (p=0.673 n=8+9)

name                             old allocs/op  new allocs/op  delta
IndexBackfill/100_rows-10           43.5k ±11%     41.4k ±10%    ~     (p=0.165 n=10+10)
IndexBackfill/10,000_rows-10         136k ± 1%      138k ± 3%    ~     (p=0.133 n=9+10)
IndexBackfill/1,000,000_rows-10     26.9M ± 1%     26.9M ± 0%    ~     (p=0.442 n=8+8)
```

annrpom commented:
rebased with the concurrency updates on master:

```
name                             old time/op    new time/op    delta
IndexBackfill/100_rows-10          21.1ms ± 2%    21.4ms ± 4%    ~     (p=0.173 n=8+10)
IndexBackfill/10,000_rows-10       38.1ms ± 3%    38.1ms ± 6%    ~     (p=0.696 n=8+10)
IndexBackfill/1,000,000_rows-10     10.9s ± 5%     10.7s ± 1%    ~     (p=0.278 n=10+9)

name                             old alloc/op   new alloc/op   delta
IndexBackfill/100_rows-10          4.52MB ± 3%    4.81MB ±20%    ~     (p=0.053 n=9+10)
IndexBackfill/10,000_rows-10       25.8MB ± 8%    24.9MB ± 3%    ~     (p=0.156 n=10+9)
IndexBackfill/1,000,000_rows-10    6.25GB ± 3%    6.21GB ± 3%    ~     (p=0.353 n=10+10)

name                             old allocs/op  new allocs/op  delta
IndexBackfill/10,000_rows-10         120k ± 2%      109k ± 2%  -9.40%  (p=0.000 n=10+10)
IndexBackfill/1,000,000_rows-10     26.9M ± 0%     25.8M ± 1%  -3.84%  (p=0.000 n=8+10)
IndexBackfill/100_rows-10           24.0k ± 1%     24.7k ± 8%    ~     (p=0.156 n=9+10)
```

we now see both 10k and 1mil improvements in allocs/op

annrpom commented:
i'm still not 100% sure why yet, but the inverted index tests were failing due to the shared key prefixes change; specifically, we need to look at a case where the inverted column (?term) has dupes:
https://github.com/cockroachdb/cockroach/blob/5a31d79468f62f7b8ae130833d3f077187552408/pkg/sql/logictest/testdata/logic_test/inverted_index#L1551
https://github.com/cockroachdb/cockroach/blob/5a31d79468f62f7b8ae130833d3f077187552408/pkg/sql/logictest/testdata/logic_test/inverted_index#L1592

when we are creating our batch of index entries, an encoding for (1, [], ...) gets created, making our entries look something like `[... (/Tenant/10/Table/116/3/[]/1/0, ...) ...]`. as soon as we get to constructing the encoded kv pair for our other `[]` value, our entries all of the sudden looks like this: 
`[... (/Tenant/10/Table/116/3/[]/5/0, ...) ... (/Tenant/10/Table/116/3/[]/5/0, ...)]` 
instead of `
[... (/Tenant/10/Table/116/3/[]/1/0, ...) ... (/Tenant/10/Table/116/3/[]/5/0, ...)]`

i tried to see if this was an instance of me assuming something is passed/call by value instead of by ref ?

annrpom committed:
sql/rowenc: reduce index key prefix calls

This patch removes redundant calls to `MakeIndexKeyPrefix` during
the construction of `IndexEntry`s by saving each first-time call in a
map that we can later lookup. Previously, we would make this call
for each row; however, as the prefix (table id + index id) for a
particular index remains the same, we do not need to do any
recomputation.

Epic: CRDB-42901
Fixes: #137798

Release note: None

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1651 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHPK8pfAvrUtwOBYt6Z:-OHPK8pfAvrUtwOBYt6_:b-dw8gem) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f62d4ab3194d3a274fa6006ae5f026a0895675/pkg/sql/rowenc/index_encoding.go#L1651)):*
> ```Go
> 	tableDesc catalog.TableDescriptor,
> 	indexes []catalog.Index,
> 	keyPrefixMap map[descpb.IndexID][]byte,
> ```

instead of making this a map, let's make this a slice. the function comment should be updated to say that the `indexes` and the `keyPrefixes` slice should both have the same ordering. i expect that to result in less overhead, since looking up in a slice is cheaper than looking up in a map.

we can create this `keyPrefixes` slice in the same place where we create the `indexesToEncode` slice:

https://github.com/cockroachdb/cockroach/blob/cca9535e048dca96d65765ac94cb51c65ac809b0/pkg/sql/backfill/backfill.go#L775-L778

https://github.com/cockroachdb/cockroach/blob/cca9535e048dca96d65765ac94cb51c65ac809b0/pkg/sql/backfill/backfill.go#L944-L965


unrelated to this PR, that second snippet makes me realize how much more expensive this is if any partial indexes are being backfilled. we have to ~reallocate~ re-populate `ib.indexesToEncode` for each row in that case.


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @yuzefovich)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1651 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHPK8pfAvrUtwOBYt6Z:-OHcdysUDBUZrdHsxauC:bhqn0sv) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f62d4ab3194d3a274fa6006ae5f026a0895675/pkg/sql/rowenc/index_encoding.go#L1651)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of making this a map, let's make this a slice. the function comment should be updated to say that the `indexes` and the `keyPrefixes` slice should both have the same ordering. i expect that to result in less overhead, since looking up in a slice is cheaper than looking up in a map.

we can create this `keyPrefixes` slice in the same place where we create the `indexesToEncode` slice:

https://github.com/cockroachdb/cockroach/blob/cca9535e048dca96d65765ac94cb51c65ac809b0/pkg/sql/backfill/backfill.go#L775-L778

https://github.com/cockroachdb/cockroach/blob/cca9535e048dca96d65765ac94cb51c65ac809b0/pkg/sql/backfill/backfill.go#L944-L965


unrelated to this PR, that second snippet makes me realize how much more expensive this is if any partial indexes are being backfilled. we have to reallocate `ib.indexesToEncode` for each row in that case.
</blockquote></details>

ok done; can you check me on this change i made:
https://github.com/cockroachdb/cockroach/blob/774bc224eb144d9b1dde61c17c8e055da9e61780/pkg/sql/backfill/backfill.go#L970-L972

my thought was that unless we clear out these key prefixes, our `indexes` and `keyPrefixes` slice are not guaranteed to maintain the same ordering as eachother (since in the case of partial indexes, we only include the index if the pred evals to true for that row). though this would mean no improvement in the case of any partial indexes being backfilled.  maybe there is another way to handle this case with slices, wdyt


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @yuzefovich)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1651 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHPK8pfAvrUtwOBYt6Z:-OHdFWw8578eGPUCRgNx:b1ks9lx) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f62d4ab3194d3a274fa6006ae5f026a0895675/pkg/sql/rowenc/index_encoding.go#L1651)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

ok done; can you check me on this change i made:
https://github.com/cockroachdb/cockroach/blob/774bc224eb144d9b1dde61c17c8e055da9e61780/pkg/sql/backfill/backfill.go#L970-L972

my thought was that unless we clear out these key prefixes, our `indexes` and `keyPrefixes` slice are not guaranteed to maintain the same ordering as eachother (since in the case of partial indexes, we only include the index if the pred evals to true for that row). though this would mean no improvement in the case of any partial indexes being backfilled.  maybe there is another way to handle this case with slices, wdyt
</blockquote></details>

i believe the logic works, but IMO it's hard to understand the code for two reasons:
- making the computation lazy makes it harder to understand which parts of the code will actually compute the value.
- modifying function parameters makes it harder to reason about what the function's inputs and outputs are.

i left a suggestion above
___
*[`pkg/sql/backfill/backfill.go` line 780 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHdFfbn8gIvG7eI88nw:-OHdFfbn8gIvG7eI88nx:b-guids) ([raw file](https://github.com/cockroachdb/cockroach/blob/774bc224eb144d9b1dde61c17c8e055da9e61780/pkg/sql/backfill/backfill.go#L780)):*
> ```Go
> 	// reset in BuildIndexEntriesChunk for every row added.
> 	ib.indexesToEncode = ib.added
> 	ib.keyPrefixes = make([][]byte, len(ib.added))
> ```

instead of computing the prefixes lazily inside of `EncodeSecondaryIndexes`, they can all just be computed here with an additional loop.
___
*[`pkg/sql/backfill/backfill.go` line 970 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHdFBNyC975gTT05F3Q:-OHdFBNyC975gTT05F3R:b-m6yoip) ([raw file](https://github.com/cockroachdb/cockroach/blob/774bc224eb144d9b1dde61c17c8e055da9e61780/pkg/sql/backfill/backfill.go#L970)):*
> ```Go
> 				if val == tree.DBoolTrue {
> 					ib.indexesToEncode = append(ib.indexesToEncode, idx)
> 				} else {
> ```

IMO this would be easier to understand if the `keyPrefixes` logic mirrored the `indexesToEncode` logic:
```
		if len(ib.predicates) > 0 {
			ib.indexesToEncode = ib.indexesToEncode[:0]
			ib.keyPrefixes = ib.keyPrefixes[:0]
			for i, idx := range ib.added {
				if !idx.IsPartial() {
					// If the index is not a partial index, all rows should have
					// an entry.
					ib.indexesToEncode = append(ib.indexesToEncode, idx)
					ib.keyPrefixes = append(ib.keyPrefixes, MakeIndexKeyPrefix(...))
					continue
				}

				// If the index is a partial index, only include it if the
				// predicate expression evaluates to true.
				texpr := ib.predicates[idx.GetID()]

				val, err := eval.Expr(ctx, ib.evalCtx, texpr)
				if err != nil {
					return nil, nil, memUsedPerChunk, err
				}

				if val == tree.DBoolTrue {
					ib.indexesToEncode = append(ib.indexesToEncode, idx)
					ib.keyPrefixes = append(ib.indexesToEncode, MakeIndexKeyPrefix(...))
				}
			}
		}
```

we can worry about optimizing this for partial indexes later (for example, by saving all of the prefixes somewhere so we can avoid calling `MakeIndexKeyPrefix` for each row).


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/backfill/backfill.go` line 780 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHdFfbn8gIvG7eI88nw:-OHmclMB6HDCReomPof9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/774bc224eb144d9b1dde61c17c8e055da9e61780/pkg/sql/backfill/backfill.go#L780)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of computing the prefixes lazily inside of `EncodeSecondaryIndexes`, they can all just be computed here with an additional loop.
</blockquote></details>

done
___
*[`pkg/sql/backfill/backfill.go` line 970 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHdFBNyC975gTT05F3Q:-OHmh-yw4xGc6yIna7S_:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/774bc224eb144d9b1dde61c17c8e055da9e61780/pkg/sql/backfill/backfill.go#L970)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

IMO this would be easier to understand if the `keyPrefixes` logic mirrored the `indexesToEncode` logic:
```
		if len(ib.predicates) > 0 {
			ib.indexesToEncode = ib.indexesToEncode[:0]
			ib.keyPrefixes = ib.keyPrefixes[:0]
			for i, idx := range ib.added {
				if !idx.IsPartial() {
					// If the index is not a partial index, all rows should have
					// an entry.
					ib.indexesToEncode = append(ib.indexesToEncode, idx)
					ib.keyPrefixes = append(ib.keyPrefixes, MakeIndexKeyPrefix(...))
					continue
				}

				// If the index is a partial index, only include it if the
				// predicate expression evaluates to true.
				texpr := ib.predicates[idx.GetID()]

				val, err := eval.Expr(ctx, ib.evalCtx, texpr)
				if err != nil {
					return nil, nil, memUsedPerChunk, err
				}

				if val == tree.DBoolTrue {
					ib.indexesToEncode = append(ib.indexesToEncode, idx)
					ib.keyPrefixes = append(ib.indexesToEncode, MakeIndexKeyPrefix(...))
				}
			}
		}
```

we can worry about optimizing this for partial indexes later (for example, by saving all of the prefixes somewhere so we can avoid calling `MakeIndexKeyPrefix` for each row).
</blockquote></details>

done
___
*[`pkg/sql/rowenc/index_encoding.go` line 1651 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHPK8pfAvrUtwOBYt6Z:-OHmdBaz398lYCgepNFu:bjpd4g1) ([raw file](https://github.com/cockroachdb/cockroach/blob/57f62d4ab3194d3a274fa6006ae5f026a0895675/pkg/sql/rowenc/index_encoding.go#L1651)):*
> it's hard to understand the code for two reasons

that's fair


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1667 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHnQPzt9wAot8h23Ztw:-OHnQPzt9wAot8h23Ztx:b245e61) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dec4f9995704013910f0589819a4e02573b246a/pkg/sql/rowenc/index_encoding.go#L1667)):*
> ```Go
> 	for i, idx := range indexes {
> 		keyPrefix := keyPrefixes[i]
> 		if keyPrefix == nil {
> ```

in what situation can `keyPrefixes[i]` be nil now that we are computing the prefixes up front?


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1667 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHnQPzt9wAot8h23Ztw:-OHnRCMV00Vwif2vVSwG:b-qbj6la) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dec4f9995704013910f0589819a4e02573b246a/pkg/sql/rowenc/index_encoding.go#L1667)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

in what situation can `keyPrefixes[i]` be nil now that we are computing the prefixes up front?
</blockquote></details>

done; this was my bad -- overlooked this change


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-:-OHniKOp8eqrdgI0-wh8:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/backfill/backfill.go` line 771 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHni9Xy2bKZ2HGhok-l:-OHni9Xy2bKZ2HGhok-m:b1i7v49) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3d5a9b9b1a8332aa95ddac57f1cb0bd50c34db2/pkg/sql/backfill/backfill.go#L771)):*
> ```Go
> 	colExprs map[descpb.ColumnID]tree.TypedExpr,
> 	mon *mon.BytesMonitor,
> 	descID descpb.ID,
> ```

just realized something as an optional nit: we can avoid needing to pass in the `descID` if we instead compute the key prefixes here: https://github.com/cockroachdb/cockroach/blob/33e120c6572ed5bc38d07a4c4e7f0cd46257deac/pkg/sql/backfill/backfill.go#L756


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/backfill/backfill.go` line 771 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHni9Xy2bKZ2HGhok-l:-OHrrQWP15cd6RkWkpm8:b-vnzfo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3d5a9b9b1a8332aa95ddac57f1cb0bd50c34db2/pkg/sql/backfill/backfill.go#L771)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

just realized something as an optional nit: we can avoid needing to pass in the `descID` if we instead compute the key prefixes here: https://github.com/cockroachdb/cockroach/blob/33e120c6572ed5bc38d07a4c4e7f0cd46257deac/pkg/sql/backfill/backfill.go#L756
</blockquote></details>

fair point -- the benefit of keeping this in `init` is that we don't do work before some fail states https://github.com/cockroachdb/cockroach/blob/893ca74c1b58d84f697fc0c7e724ee48569cd3b6/pkg/sql/backfill/backfill.go#L651-L654 (like https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/backfill/index_backfiller_cols.go#L103-L105)


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/backfill/backfill.go` line 771 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHni9Xy2bKZ2HGhok-l:-OHs6Cz3BT9XL4tpM3Sd:b-uwszn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3803f62fca09bb5df4ca306e995f5068b52da86/pkg/sql/backfill/backfill.go#L771)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

fair point -- the benefit of keeping this in `init` is that we don't do work before some fail states https://github.com/cockroachdb/cockroach/blob/893ca74c1b58d84f697fc0c7e724ee48569cd3b6/pkg/sql/backfill/backfill.go#L651-L654 (like https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/backfill/index_backfiller_cols.go#L103-L105)
</blockquote></details>

i see. encoding a constant number of index prefixes is a pretty tiny amount of work though, so i don't think we're saving much on overhead.

it's just a code organization preference, so i'll leave this up to you.
___
*[`pkg/sql/backfill/backfill.go` line 789 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHs6XVPAhFPhXs5CK5X:-OHs6XVPAhFPhXs5CK5Y:b-7aoi12) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3803f62fca09bb5df4ca306e995f5068b52da86/pkg/sql/backfill/backfill.go#L789)):*
> ```Go
> 	if len(ib.predicates) > 0 {
> 		ib.indexesToEncode = make([]catalog.Index, 0, len(ib.added))
> 		ib.keyPrefixes = make([][]byte, 0, len(ib.added))
> ```

another optional comment: since `keyPrefixes` hasn't been initialized before, we could move this `make` call to be above the for loop, that way the slice only ever gets allocated once. again, this is a fairly trivial amount of allocation savings so it's more about style.


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/backfill/backfill.go` line 771 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHni9Xy2bKZ2HGhok-l:-OHt4Ezf5VqiJLe_HmdA:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3803f62fca09bb5df4ca306e995f5068b52da86/pkg/sql/backfill/backfill.go#L771)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i see. encoding a constant number of index prefixes is a pretty tiny amount of work though, so i don't think we're saving much on overhead.

it's just a code organization preference, so i'll leave this up to you.
</blockquote></details>

done
___
*[`pkg/sql/backfill/backfill.go` line 789 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHs6XVPAhFPhXs5CK5X:-OHsyXfG2Lbes3YHUyhy:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3803f62fca09bb5df4ca306e995f5068b52da86/pkg/sql/backfill/backfill.go#L789)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

another optional comment: since `keyPrefixes` hasn't been initialized before, we could move this `make` call to be above the for loop, that way the slice only ever gets allocated once. again, this is a fairly trivial amount of allocation savings so it's more about style.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-:-OHt6o0lEpgBLxfJNynD:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1668 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHt6i4v0vHs4sFpuUJt:-OHt6i4v0vHs4sFpuUJu:b-2qx89w) ([raw file](https://github.com/cockroachdb/cockroach/blob/9152c31aac59309e1932faa3ae02259849433d8f/pkg/sql/rowenc/index_encoding.go#L1668)):*
> ```Go
> 	for i, idx := range indexes {
> 		keyPrefix := keyPrefixes[i]
> 		if idx.GetType() == idxtype.INVERTED {
> ```

could you add a TODO comment that says what failed when we didn't have the inverted index exception?


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139256)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1670 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/139256#-OHx-6yQ9O8WbWCHBC-V:-OHx-6yQ9O8WbWCHBC-W:b-efqjwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1697d7522134854a6c1deec195c3dcedbe61f42c/pkg/sql/rowenc/index_encoding.go#L1670)):*
> ```Go
> 		// TODO(annie): For now, we recompute the key prefix of inverted indexes. This is because index
> 		// keys with multiple associated values somehow get encoded into the same kv pair when using
> 		// our precomputed key prefix.
> ```

super nit: it would be helpful to future debuggers if this comment could include the name of a test that was failing


<!-- Sent from Reviewable.io -->

annrpom commented:
TFTR! ('-')7

bors r+
